### ISSUE TITLE
Trying to get catboost to use GPU from R

### ISSUE THREAD

**Issue by timchurches** [catboost/catboost#481]
Problem: Unable to get catboost to use the GPU. R 3.5.1 coiled from source on Linux. CatBoost installed from source using `devtools::install_github('catboost/catboost', subdir = 'catboost/R-package')`, package compiles without errors using clang. Passes all tests, works fine with task_type='CPU'. However, with task_type='GPU', get this error:

```
fit_params <- list(iterations = 1000,
                   thread_count = 10,
                   loss_function = 'Logloss',
                   ignored_features = c(4,9),
                   border_count = 32,
                   depth = 5,
                   learning_rate = 0.03,
                   l2_leaf_reg = 3.5,
                   # border = 0.5,
                   task_type = "GPU",
                   # devices = 1,
                   train_dir = 'train_dir')
model <- catboost.train(pool, test_pool, fit_params)

Error in catboost.train(pool, test_pool, fit_params) :    catboost/libs/train_lib/train_model.cpp:787: Can't load GPU learning library. Module was not compiled or driver  is incompatible with package. Please install latest NVDIA driver and check again
```

Tried upgrading to CUDA v10, re-compile catboost, same result.

catboost version: 0.10.2 (latest from GitHub master)
Operating System: Amazon Linux Deep Learning AMI v14 (RedHat/CentOS-based) 
CPU: x86_64 (AWS p3.2xlarge instance)
# GPU: NVIDIA V100 (one only)

```
$ nvidia-smi -q | head

==============NVSMI LOG==============

Timestamp                           : Tue Sep 25 02:05:57 2018
Driver Version                      : 396.44

Attached GPUs                       : 1
GPU 00000000:00:1E.0
    Product Name                    : Tesla V100-SXM2-16GB
    Product Brand                   : Tesla
```

**Comment by timchurches** [comment]
Ah, this may be the problem:

```
$ nvcc --version
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2017 NVIDIA Corporation
Built on Fri_Sep__1_21:08:03_CDT_2017
Cuda compilation tools, release 9.0, V9.0.176
```

Does catboost need CUDA 9.1 ?

**Comment by dbakshee** [comment]
Hi, this is an issue in the R-package/src/Makefile. Command `ya make` should be given option `-DCUDA_ROOT=...` explicitly. I'll fix this soon. 
Thanks
Dima

**Comment by timchurches** [comment]
Looked at this, but I can't work out how to add it to ya make... the Yandex make system looks interesting, but not clear exactly how it works...

When this is working, we'll publish and maintain a community AWS AMI based on the AWS Deep learning AMI which includes R, RStudio Server, and CatBoost compiled to use GPUs (plus tensorflow and keras for R etc), for deployment on AWS GPUs-enbaled instances (particularly the p3 instances with Tesla V100 GPUs attached - they are really fast).

**Comment by timchurches** [comment]
> Sorry for my question, but why prebuilt R package don't work for you?

Um, good question! I completely overlooked the binary release for v0.10...

So, that installed perfectly, and it works! Definitely using the GPU (as shown by `nvidia-smi pmon`). However, on an AWS p3.2xlarge instance (8 vCPUs, 1 x Tesla V100 GPU, 61GB RAM), the GPU mode is quite a bit slower than the CPU mode, although it seems to give slightly better results, mostly (it isn't deterministic...). Anyway, many thanks for your assistance!

```
> # From file
> pool_path <- system.file("extdata", "adult_train.1000", package = "catboost")
> test_pool_path <- system.file("extdata", "adult_test.1000", package = "catboost")
> cd_path <- system.file("extdata", "adult.cd", package = "catboost")
> pool <- catboost.load_pool(pool_path, column_description = cd_path)
> test_pool <- catboost.load_pool(test_pool_path, column_description = cd_path)
> 
> fit_params <- list(iterations = 2000,
+                    thread_count = 10,
+                    loss_function = 'Logloss',
+                    ignored_features = c(4,9),
+                    border_count = 32,
+                    depth = 5,
+                    learning_rate = 0.03,
+                    l2_leaf_reg = 3.5,
+                    task_type='CPU',
+                    # border = 0.5,
+                    train_dir = 'train_dir',
+                    metric_period=1000)
> model <- catboost.train(pool, test_pool, fit_params)
0:	learn: 0.6734723	test: 0.6739090	best: 0.6739090 (0)	total: 8.5ms	remaining: 17s
1000:	learn: 0.2224211	test: 0.4029066	best: 0.4029066 (1000)	total: 7.69s	remaining: 7.67s
1999:	learn: 0.1634618	test: 0.4223962	best: 0.4029066 (1000)	total: 15.5s	remaining: 0us

bestTest = 0.4029065638
bestIteration = 1000

Shrink model to first 1001 iterations.
> 
> fit_params <- list(iterations = 2000,
+                    thread_count = 10,
+                    loss_function = 'Logloss',
+                    ignored_features = c(4,9),
+                    border_count = 32,
+                    depth = 5,
+                    learning_rate = 0.03,
+                    l2_leaf_reg = 3.5,
+                    task_type='GPU',
+                    # border = 0.5,
+                    train_dir = 'train_dir',
+                    metric_period=1000)
> model <- catboost.train(pool, test_pool, fit_params)
0:	learn: 0.6761206	test: 0.6772648	best: 0.6772648 (0)	total: 26.7ms	remaining: 53.4s
1000:	learn: 0.1886293	test: 0.4117968	best: 0.4117968 (1000)	total: 23.5s	remaining: 23.5s
1999:	learn: 0.1321974	test: 0.4386451	best: 0.4117968 (1000)	total: 47s	remaining: 0us
bestTest = 0.4117967529
bestIteration = 1000
Shrink model to first 1001 iterations.
```

**Comment by kizill** [comment]
@timchurches adult_train.1000 only contains 1k objects and GPU learning reveals it's incredible speed only when you are using relatively big datasets (it depends on feature layout of datasets, but as i remember gpu training starts to overtake cpu training between 100k-1mln training samples.

**Comment by timchurches** [comment]
> @timchurches adult_train.1000 only contains 1k objects and GPU learning reveals it's incredible speed only when you are using relatively big datasets (it depends on feature layout of datasets, but as i remember gpu training starts to overtake cpu training between 100k-1mln training samples.

OK, thanks. Our dataset contains 1.3M rows, so we are training on nearly 1M rows. I'll report what sort of speed-up we get when we try that tomorrow. Thanks again, CatBoost rocks!

**Comment by annaveronika** [comment]
GPU training starts to overtake with around 20-30k samples. With 50k + Volta you will have around 10 times speedup, with 1kk samples you will have around 40-50 times speedup.
About quality - it is the same quality, so it is safe to use GPU. We train most of our production models in Yandex on GPUs.

**Comment by timchurches** [comment]
Thanks @annaveronika. So far we are seeing slightly better predictive performance with the GPU - maybe the permutations on the categories are done differently on the GPU? But yeah, it is now **really** fast to train on our full dataset. We will write up our experiences and findings with GPU-enabled CatBoost in a blog post next week. We're working with hospital clinical data (intensive care unit data at the moment), which is a mixture of categorical, ordinal and continuous features, much of them missing... which is why CatBoost is such a godsend - it makes working with such data so much easier, as well as providing better predictive performance than everything else (we are benchmarking against XGBoost).

**Comment by annaveronika** [comment]
There are small algorithmic differences between CPU and GPU, like using float instead of double on GPU, or a some difference in gradient walker. But I think main difference is because there are different random permutations (CPU and GPU use different seed for that).

I'm very happy to hear that it worked well for you! And we will be happy if you write this blogpost.

**Comment by timchurches** [comment]
Quick update: CatBoost for R now running well on GPU using the binary install of v0.10.0 as suggested. Running on an AWS p3.2xlarge instance (8 CPU cores, 61GB memory, one Nvidia Tesla V100 GPU with 16GB memory) running Amazon Linux. Training time for about 700k row with about 40 features, mixture of categorical and continuous, is about 9 hrs using the CPUs (all 8 cores), and about 15 minutes using the GPU, so close to a 40 times speed-up as promised! Fantastic! Predictive performance seems to be a little better on GPU, but now it is feasible for us to properly optimise the hyperparameters, so we expect to do quite a bit better on the GPU (we're already doing much better than anything reported in the scientific literature so far)  I'll write this up in a blog post and let you know next week. Many thanks to the CatBoost team from the ICU modelling team at the Centre for Big Data Research in Health in the Faculty of Medicine at the University of NSW, Sydney.

**Comment by sab** [comment]
Hey @timchurches !

Did you have a chance to write a blog post about your results with CatBoost? We'll be happy to share the link through CatBoost twitter account.

Furthermore we'll appreciate if you'll drop a line about your case into https://github.com/catboost/catboost/blob/master/AWESOME.md

**Comment by sab** [comment]
@timchurches Great! Will wait for your blog post!

**Comment by annaveronika** [comment]
@timchurches Could you tell us more details about your CatBoost usage? What is the problem that you are solving?
